alm/src/utils/contract.ts

```diff
@@ -7,12 +7,46 @@ export const getRequiredBlockConfirmations = async (contract: Contract) => {
 
 export const getValidatorAddress = (contract: Contract) => contract.methods.validatorContract().call()
 
-export const getRequiredSignatures = async (contract: Contract) => {
-  const requiredSignatures = await contract.methods.requiredSignatures().call()
+export const getRequiredSignatures = async (contract: Contract, blockNumber: number) => {
+  const events = await contract.getPastEvents('RequiredSignaturesChanged', {
+    fromBlock: 0,
```

Comment:
Consider to have this changes applied to the Ethereum Mainnet, does it make sense to start from the block 0?

Do you think it is possible to generate a snapshot of events (not only `RequiredSignaturesChanged`) for the ALM application as part of deployment process so the application will request just delta between the snapshot and current state of the blockchain? I believe it could make the application more respondent.

Reply:
> Consider to have this changes applied to the Ethereum Mainnet, does it make sense to start from the block 0?

I thought about it, one alternative is to first call `deployedAtBlock` and then use it as the `fromBlock` parameter. But in this case, I think it would take more time to get the result because of doing one request, wait for the result, and then do the second request. That's why I have chosen to use the block `0`

> Do you think it is possible to generate a snapshot of events (not only RequiredSignaturesChanged) for the ALM application as part of deployment process so the application will request just delta between the snapshot and current state of the blockchain? I believe it could make the application more respondent.

I think it is possible. We could develop a script that will get for example the events of `RequiredSignaturesChanged`, `ValidatorAdded`, and `ValidatorRemoved`. The script could run before building the application and store the results in a file that will be included in the sources (to avoid having other external services). Then the app will get the information from these snapshots and only request the delta.

The snapshots could also include the basic information of the app such as:
- Chain Ids
- Block confirmations numbers
- Validator contracts address (and for example `deployedAtBlock` value if needed) 

Having this information will save doing a lot of requests, which will increase the response of the app. I wouldn't recommend storing information about the user's transactions because it could be a lot of data, and it will affect the loading time of the app because of the size of the snapshots.

Should we create a new issue for snapshot improvement?

Reply:
> Block confirmations numbers

BTW, this number could be changed. And there is no way to track these transactions by event, isn't it?

> Should we create a new issue for snapshot improvement?

Yes, please

Reply:
>BTW, this number could be changed. And there is no way to track these transactions by event, isn't it?

The event `RequiredBlockConfirmationChanged` is emitted when the number is changed, so I think it makes sense to also create an issue to get the block confirmations similar as we are doing for required signatures using the events

Reply:
Cool! Let's create an issue and keep in mind that these events also should be in a snapshot we discussed above.

Reply:
Created the following issues:
- For Block confirmations changes https://github.com/poanetwork/tokenbridge/issues/369
- For snapshot https://github.com/poanetwork/tokenbridge/issues/370